test/randtilegen.c

```diff
@@ -0,0 +1,264 @@
+			exit(0);
+		}
+		randBits |= (uint32_t)data << randCount;
+		randCount += 8;
+		fputc(data, rngRecorder);
+	}
+
+	uint32_t result = randBits & (((uint32_t)1 << count) - 1);
+	randBits >>= count;
+	randCount -= count;
+	return result;
+}
+
+/**
+ * Flush any remaining bits in the RNG storage
+ */
+static void flushRng(void) {
+	randCount = 0;
+	randBits = 0;
+}
+
+/**
+ * Expand a 5-bit color component to 8 bits with minimal bias
+ */
+static uint8_t _5to8(uint8_t five) {
+	return five << 3 | five >> 2;
+}
+
+static void generate_random_image(png_structp png, png_infop pngInfo) {
+#define NB_TILES 10 * 10
+	struct {
+		unsigned char palette;
+		unsigned char nbColors;
+	} attributes[NB_TILES];
+	uint8_t tileData[NB_TILES][8][8];
+	// These two are in tiles, not pixels, and in range [3; 10], hence `NB_TILES` above
+	// Both width and height are 4-bit values, so nbTiles is 8-bit (OK!)
+	uint8_t const width = getRandomBits(3) + 3, height = getRandomBits(3) + 3,
+	              nbTiles = width * height;
+
+	for (uint8_t p = 0; p < nbTiles; p++) {
+		uint8_t pal;
+		do {
+			pal = getRandomBits(5);
+		} while (pal == 0 || (pal > 29));
+		attributes[p].palette = 2 * pal + getRandomBits(1);
+		// Population count (nb of bits set), the simple way
+		static uint8_t const popcount[] = {1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4,
+		                                   1, 2, 2, 3, 2, 3, 3, 4, 2, 3, 3, 4, 3, 4};
+		attributes[p].nbColors = popcount[pal - 1];
+		// Handle single-color tiles the simple way, without trying to pull more random bits
+		if (attributes[p].nbColors < 2) {
+			memset(tileData[p], 0, sizeof(tileData[p]));
+			continue;
+		}
+		uint8_t index, total;
+		for (index = 0, total = 0; index < p; index++) {
+			if (attributes[index].nbColors == attributes[p].nbColors) {
+				total++;
+			}
+		}
+		// index == p at exit
+		if (total) {
+			index = getRandomBits(8);
+			if (index < total) {
+				total = index + 1;
+				for (index = 0; total; index++) {
+					if (attributes[index].nbColors == attributes[p].nbColors) {
+						total--;
+					}
+					if (!total) {
+						index--;
+					}
+				}
+			} else {
+				index = p;
+			}
+		}
+		if (index != p) {
+			unsigned rotation = getRandomBits(2);
+			for (uint8_t y = 0; y < 8; y++) {
+				for (uint8_t x = 0; x < 8; x++) {
+					tileData[p][y][x] =
+					    tileData[index][y ^ ((rotation & 2) ? 7 : 0)][x ^ ((rotation & 1) ? 7 : 0)];
+				}
+			}
+		} else {
+			switch (attributes[p].nbColors) {
+			case 2:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++)
+						tileData[p][y][x] = getRandomBits(1);
+				break;
+			case 4:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++)
+						tileData[p][y][x] = getRandomBits(2);
+				break;
+			default:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++) {
+						do {
+							index = getRandomBits(2);
+						} while (index == 3);
+						tileData[p][y][x] = index;
+					}
```

Comment:
A multi-line `for` body should absolutely have braces, even if the multi-lines are technically a single block statement.
```suggestion
				for (uint8_t y = 0; y < 8; y++) {
					for (uint8_t x = 0; x < 8; x++) {
						do {
							index = getRandomBits(2);
						} while (index == 3);
						tileData[p][y][x] = index;
					}
				}
```

Reply:
Oops, that one escaped me :P